### ISSUE TITLE
Design a flatbuffers based map format

### ISSUE THREAD

**Issue by djarek** [otland/forgottenserver#2206]
# Design goals
- achieve faster map loading time
- keep the map size within the order of magnitude of OTBM
- be compatible with #2139

# Other requirements
- provide conversion tools
- implement support for the new format in map editors ?

**Comment by ranisalt** [comment]
Can you provide an OTBM format specification for reference? I think what is currently specified in the source code is very lacking and too superficial about what can be contained in an OTBM file.

**Comment by djarek** [comment]
This is a very generic spec. Too bad the original author didn't write an RFC :D.
```
Escape = 0xFD
Start = 0xFE
End = 0xFF
NodeTag = Escape | Start | End
OTBM_ROOTV1 = 0x01
NodeType = OTBM_ROOTV1 | 0x02 | ... | 0x10
EscapedByte = (0xFD 0xFD) | (0xFD 0xFE) | (0xFD 0xFF)
UnescapedByte = 0x00 | 0x01 | ... | 0xFC
PropertyStream = {EscapedByte | UnescapedByte}
Node = Start NodeType [PropertyStream] {Node} End
Identifier = "OTBM" | "OTBI"
RootHeaderVersion = u32le
RootHeaderWidth = u16le
RootHeaderHeight = u16le
RootHeaderMajorItemsVersion = u32le
RootHeaderMinorItemsVersion = u32le
RootHeader = RootHeaderVersion RootHeaderWidth RootHeaderHeight RootHeaderMajorItemsVersion RootHeaderMinorItemsVersion
RootNode = Start OTBM_ROOTV1 {Node} End
OTBFile = Identifier RootNode 
```

**Comment by conde2** [comment]
Wouldn't worth rework OTB aswell when doing this changes ? It could totally be done in XML/LUA or anything more friendly than what we have now.

**Comment by djarek** [comment]
@conde2 
If you mean the items file, I'm not sure why it's in binary format now, if there are no size issue, then it could be merged with items.xml. As for xml maps? Nope, not going to happen.

**Comment by conde2** [comment]
Yes, I was talking about the items.otb, it would be awesome have it  merged with items.XML.
About the map I know it is not viable doing it in XML, i was strictly speaking about items.

**Comment by ranisalt** [comment]
Expect items.xml to grow orders of magnitude in size since binary formats can store much more info on a single byte compared to text formats.

**Comment by Kamenuvol** [comment]
I do not like items.xml format either. It's so repetitive. Why not use custom tags instead of attributes? 
An items.lua would be smaller size including everything otb and xml have and it's easier to create custom attributes, except the ones we use binary shifting to join some of them.

**Comment by djarek** [comment]
TBH I don't consider a size increase of 3x or more to be acceptable. If migrating item information fully to XML is not feasible then we should consider using a simillar protobuf format for all item information as well.

**Comment by Nu77** [comment]
OTB is a pain in the ass. Theres no cross-platform tools to handle it properly.
About size maybe a module file like this could be useful:
```xml
<file name="781" path="items781.xml" /> // equips items
<file name="854" path="items854.xml" /> // unmovable items
<file name="960" path="items960.xml" /> // quests items
```
I've used tibia versions as example since we use it as standard.

**Comment by djarek** [comment]
First draft for OTFM an OTFI file format schemas:
https://github.com/otland/forgottenserver/commit/d01e041409ab3877c5c71fd066bfa7577b73249a

Some open design questions:
- should sectors have explicit sizes?
- can we make lazy loading sectors easier?
- is it feasible to use the Item schema to store items in the database?

**Comment by conde2** [comment]
should sectors have explicit sizes?
I think yes, it makes loading more consistent, faster (no need to calculate sectors) and I can't see drawbacks on this.

can we make lazy loading sectors easier?
Easier in which way ?

is it feasible to use the Item schema to store items in the database?
How it will work with custom attributes ?

**Comment by ranisalt** [comment]
I think you are working too much on porting the current format rather than designing a new format from the ground up, e.g.

- I wouldn't store house information on a tile, but on a separate section or even separate file
- Items version shouldn't be related with map once you use a flatbuffers format
- Towns should have their own XML/Lua file
- Map version also wouldn't be relevant with flatbuffers

FB has the advantage of being pluggable should we need any other info in the future, and itself would reduce file size by a lot.

Also, new map format and new item format should be designed separately.
